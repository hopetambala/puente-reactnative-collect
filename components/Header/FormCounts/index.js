import React, { useState } from 'react';
import { useEffect } from 'react/cjs/react.development';
import {
  View
} from 'react-native';
import { Button, Text } from 'react-native-paper';
import { Spinner } from 'native-base';

import { getData } from '../../../modules/async-storage';
import { countService } from '../../../services/parse/crud';
import styles from '../index.styles';
import { theme } from '../../../modules/theme';

const FormCounts = ({ setShowCounts }) => {
  const [surveyCount, setSurveyCount] = useState(0);
  const [envHealthCount, setEnvHealthCount] = useState(0);
  const [vitalsCount, setVitalsCount] = useState(0);
  const [customCount, setCustomCount] = useState(0);
  const [userName, setUserName] = useState(' ');
  const [surveyQueryDone, setSurveyQueryDone] = useState(false)
  const [envHealthQueryDone, setEnvHealthQueryDone] = useState(false)
  const [vitalsQueryDone, setVitalsQueryDone] = useState(false)
  const [customQueryDone, setCustomQueryDone] = useState(false)

  useEffect(() => {
    getData('currentUser').then((user) => {
      const username = `${user.firstname || ''} ${user.lastname || ''}`;
      setUserName(username);
    });
  }, []);

  useEffect(() => {
    const postParamsSurvey = {
      ParseClass: 'SurveyData',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    countService(postParamsSurvey).then((surveyCounts) => {
      setSurveyCount(surveyCounts);
      setSurveyQueryDone(true);
    });
  }, [userName]);

  useEffect(() => {
    const postParamsEnvHealth = {
      ParseClass: 'HistoryEnvironmentalHealth',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    countService(postParamsEnvHealth).then((envHealthCounts) => {
      setEnvHealthCount(envHealthCounts);
      setEnvHealthQueryDone(true)
    });
  }, [userName]);

  useEffect(() => {
    const postParamsVitals = {
      ParseClass: 'Vitals',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    countService(postParamsVitals).then((vitalsCounts) => {
      setVitalsCount(vitalsCounts);
      setVitalsQueryDone(true)
    });
  }, [userName]);

  useEffect(() => {
    const postParamsCustomForms = {
      ParseClass: 'FormResults',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    countService(postParamsCustomForms).then((customCounts) => {
      setCustomCount(customCounts);
      setCustomQueryDone(true);
    });
  }, [userName]);

  return (
    <View>
      <Text style={styles.headerFormText}>Surveys Collected</Text>
      <View style={styles.horizontalLineGray} />
      {surveyQueryDone && envHealthQueryDone && vitalsQueryDone && customQueryDone ? (
        <View>
          <View style={styles.countContainer}>
            <Text style={styles.label}>ID Forms</Text>
            <Text style={styles.count}>{surveyCount}</Text>
          </View>
          <View style={styles.horizontalLineGray} />
          <View style={styles.countContainer}>
            <Text style={styles.label}>Environmental Health Forms</Text>
            <Text style={styles.count}>{envHealthCount}</Text>
          </View>
          <View style={styles.horizontalLineGray} />
          <View style={styles.countContainer}>
            <Text style={styles.label}>Vitals Forms</Text>
            <Text style={styles.count}>{vitalsCount}</Text>
          </View>
          <View style={styles.horizontalLineGray} />
          <View style={styles.countContainer}>
            <Text style={styles.label}>Custom Forms</Text>
            <Text style={styles.count}>{customCount}</Text>
          </View>
          <View style={styles.horizontalLineGray} />
        </View>
      ) : (
        <Spinner color={theme.colors.primary} />
      )}
      <Button onPress={() => setShowCounts(false)}>Back</Button>
    </View>
  );
};

export default FormCounts;
