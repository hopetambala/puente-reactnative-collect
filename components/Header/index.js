import React from 'react';
import { View } from 'react-native';
import { Text, IconButton } from 'react-native-paper';

import styles from './index.styles';

const Header = () => {
  const { header, headerText, headerIcon } = styles;
  return (
    <View style={header}>
      <Text style={headerText}>PUENTE</Text>
      <View style={headerIcon}>
        <IconButton
          icon="account"
          color={headerIcon.color}
          size={30}
        />
      </View>
      <View style={headerIcon}>
        <IconButton
          icon="settings"
          color={headerIcon.color}
          size={30}
        />
      </View>
    </View>
  );
};

export default Header;
