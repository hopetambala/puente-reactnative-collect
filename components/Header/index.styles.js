import { StyleSheet } from 'react-native';
import { theme } from '../../modules/theme';

const borderRadius = 20;

const { accent, black } = theme.colors;

const styles = StyleSheet.create({
  header: {
    height: 80, // equivalent to flex: 0.2,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-evenly',
    borderBottomRightRadius: borderRadius,
    borderBottomLeftRadius: borderRadius,
    backgroundColor: accent
  },
  headerIcon: {
    borderRadius: 30,
    color: black
  },
  headerText: {
    fontSize: 30,
    fontWeight: 'bold',
    color: black,
    flex: 0.7
  },

});

export default styles;
