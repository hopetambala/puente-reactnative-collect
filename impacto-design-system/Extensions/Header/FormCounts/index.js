import { countService } from '@app/services/parse/crud';
import { getData } from '@modules/async-storage';
import checkOnlineStatus from '@modules/offline';
import { theme } from '@modules/theme';
import I18n from 'i18n-js';
import { Spinner } from 'native-base';
import React, { useEffect, useState } from 'react';
import {
  View
} from 'react-native';
import { Button, Text } from 'react-native-paper';

import styles from '../index.styles';

const FormCounts = ({ setShowCounts }) => {
  const [surveyCount, setSurveyCount] = useState(0);
  const [envHealthCount, setEnvHealthCount] = useState(0);
  const [vitalsCount, setVitalsCount] = useState(0);
  const [customCount, setCustomCount] = useState(0);
  const [assetCount, setAssetCount] = useState(0);
  const [userName, setUserName] = useState(' ');
  const [error, setError] = useState(false);
  const [offline, setOffline] = useState(false);

  const [queryDone, setQueryDone] = useState(false);

  useEffect(() => {
    getData('currentUser').then((user) => {
      const username = `${user.firstname || ''} ${user.lastname || ''}`;
      setUserName(username);
    });
  }, []);

  useEffect(() => {
    const postParamsSurvey = {
      ParseClass: 'SurveyData',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    const postParamsEnvHealth = {
      ParseClass: 'HistoryEnvironmentalHealth',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    const postParamsVitals = {
      ParseClass: 'Vitals',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    const postParamsCustomForms = {
      ParseClass: 'FormResults',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };
    const postParamsAssets = {
      ParseClass: 'FormResults',
      parseColumn: 'surveyingUser',
      parseParam: userName
    };

    const idPromise = countService(postParamsSurvey);
    const envHealthPromise = countService(postParamsEnvHealth);
    const vitalsPromise = countService(postParamsVitals);
    const customFormsPromise = countService(postParamsCustomForms);
    const assetPromise = countService(postParamsAssets);

    checkOnlineStatus().then(async (connected) => {
      if (connected) {
        Promise.all([idPromise, envHealthPromise,
          vitalsPromise, customFormsPromise,
          assetPromise]).then((values) => {
          setSurveyCount(values[0]);
          setEnvHealthCount(values[1]);
          setVitalsCount(values[2]);
          setCustomCount(values[3]);
          setAssetCount(values[4]);
          setQueryDone(true);
        }, () => {
          // error - maybe inform user
          setError(true);
        });
      } else {
        setOffline(true);
      }
    });
  }, [userName]);

  return (
    <View>
      <Text style={styles.headerFormText}>{I18n.t('formCounts.surveysCollected')}</Text>
      <View style={styles.horizontalLineGray} />
      {error && (
      <View>
        <Text style={styles.label}>
          f
          {I18n.t('formCounts.error')}
        </Text>
      </View>
      )}
      {offline && (
      <View>
        <Text style={styles.label}>{I18n.t('formCounts.offline')}</Text>
      </View>
      )}
      {queryDone && (
      <View>
        <View style={styles.countContainer}>
          <Text style={styles.label}>{I18n.t('formCounts.idForms')}</Text>
          <Text style={styles.count}>{surveyCount}</Text>
        </View>
        <View style={styles.horizontalLineGray} />
        <View style={styles.countContainer}>
          <Text style={styles.label}>{I18n.t('formCounts.envHealthForms')}</Text>
          <Text style={styles.count}>{envHealthCount}</Text>
        </View>
        <View style={styles.horizontalLineGray} />
        <View style={styles.countContainer}>
          <Text style={styles.label}>{I18n.t('formCounts.vitalsForms')}</Text>
          <Text style={styles.count}>{vitalsCount}</Text>
        </View>
        <View style={styles.horizontalLineGray} />
        <View style={styles.countContainer}>
          <Text style={styles.label}>{I18n.t('formCounts.customForms')}</Text>
          <Text style={styles.count}>{customCount}</Text>
        </View>
        <View style={styles.horizontalLineGray} />
        <View style={styles.countContainer}>
          <Text style={styles.label}>{I18n.t('formCounts.assetForms')}</Text>
          <Text style={styles.count}>{assetCount}</Text>
        </View>
        <View style={styles.horizontalLineGray} />
      </View>
      )}
      {!queryDone && !error && !offline && (
      <Spinner color={theme.colors.primary} />
      )}
      <Button onPress={() => setShowCounts(false)}>{I18n.t('formCounts.back')}</Button>
    </View>
  );
};

export default FormCounts;
