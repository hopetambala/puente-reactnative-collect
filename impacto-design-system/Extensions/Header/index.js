import { OfflineContext } from "@context/offline.context";
import { getData } from "@modules/async-storage";
import handleParseError from "@modules/cached-resources/error-handling";
import I18n from "@modules/i18n";
import checkOnlineStatus from "@modules/offline";
import {
  cleanupPostedOfflineForms,
  postOfflineForms,
} from "@modules/offline/post";
import { Spinner } from "native-base";
import React, { useContext, useState } from "react";
import { View } from "react-native";
import Emoji from "react-native-emoji";
import { Button, Headline, IconButton, Text } from "react-native-paper";

import FormCounts from "./FormCounts";
import styles from "./index.styles";

const Header = ({ setSettings }) => {
  const { header, headerText, headerIcon } = styles;
  const [drawerOpen, setDrawerOpen] = useState(false);
  const [volunteerDate, setVolunteerDate] = useState("");
  const [volunteerGreeting, setVolunteerGreeting] = useState("");
  const [offlineForms, setOfflineForms] = useState(false);
  const [offlineFormCount, setOfflineFormCount] = useState(0);
  const [submission, setSubmission] = useState(null);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [showCounts, setShowCounts] = useState(false);
  const { populateResidentDataCache, isLoading: isOfflineLoading } =
    useContext(OfflineContext);

  const volunteerLength = (object) => {
    const date = new Date(object.createdAt);
    const convertedDate = date.toDateString();
    return convertedDate;
  };

  const calculateTime = (name) => {
    const today = new Date();
    const curHr = today.getHours();

    if (curHr < 12) {
      setVolunteerGreeting(
        `${I18n.t("header.goodMorning")} ${name}!` ||
          I18n.t("header.goodMorning!")
      );
    } else if (curHr < 18) {
      setVolunteerGreeting(
        `${I18n.t("header.goodAfternoon")} ${name}!` ||
          I18n.t("header.goodAfternoon!")
      );
    } else {
      setVolunteerGreeting(
        `${I18n.t("header.goodEvening")} ${name}!` ||
          I18n.t("header.goodEvening!")
      );
    }
  };

  const count = async () => {
    getData("currentUser").then((user) => {
      calculateTime(user.firstname);
      setVolunteerDate(volunteerLength(user));
    });

    const idFormCount = await getData("offlineIDForms").then((idForms) => {
      if (idForms) {
        setOfflineForms(true);
        return idForms.length;
      }
      return 0;
    });

    const supplementaryCount = await getData("offlineSupForms").then(
      (supForms) => {
        if (supForms) {
          setOfflineForms(true);
          return supForms.length;
        }
        return 0;
      }
    );

    const assetIdFormCount = await getData("offlineAssetIDForms").then(
      (assetIdForms) => {
        if (assetIdForms) {
          setOfflineForms(true);
          return assetIdForms.length;
        }
        return 0;
      }
    );

    const assetSupForms = await getData("offlineAssetSupForms").then(
      (assetSuppForms) => {
        if (assetSuppForms) {
          setOfflineForms(true);
          return assetSuppForms.length;
        }
        return 0;
      }
    );

    const allFormOfflineCount =
      idFormCount + supplementaryCount + assetIdFormCount + assetSupForms;

    setOfflineFormCount(allFormOfflineCount);

    setOfflineForms(allFormOfflineCount > 0);

    setDrawerOpen(!drawerOpen);
  };

  const upload = async () => {
    setIsSubmitting(true);
    const offlineRecords = await postOfflineForms().catch((error) =>
      handleParseError(error, postOfflineForms).then(async (records) => {
        const { status } = records;

        if (status === "Error") {
          setIsSubmitting(false);
          setSubmission(false);
          return;
        }
        setSubmission(true);
        setIsSubmitting(false);
        await cleanupPostedOfflineForms();
      })
    );

    const { status } = offlineRecords;
    if (status === "Error") {
      setIsSubmitting(false);
      setSubmission(false);
      return;
    }

    setSubmission(true);
    setIsSubmitting(false);
    await cleanupPostedOfflineForms();
  };

  const cacheOfflineData = async () =>
    checkOnlineStatus().then(async (connected) => {
      if (connected) await populateResidentDataCache();
    });

  const navToSettings = () => {
    setDrawerOpen(false);
    setSettings(true);
  };

  const navToCounts = () => {
    setShowCounts(true);
  };

  return (
    <View style={styles.container}>
      <View style={header}>
        <Text style={headerText}>{I18n.t("header.puente")}</Text>
        <View style={headerIcon}>
          <IconButton color={headerIcon.color} size={30} />
        </View>
        <View style={headerIcon}>
          <IconButton
            icon="tune"
            color={headerIcon.color}
            size={30}
            onPress={navToSettings}
          />
        </View>
      </View>
      {drawerOpen && (
        <View>
          {!showCounts ? (
            <View>
              <Headline style={styles.calculationText}>
                {volunteerGreeting}
                <Emoji name="coffee" />
              </Headline>
              <View style={{ flexDirection: "row", alignSelf: "center" }}>
                <Text style={styles.calculationText}>{`${I18n.t(
                  "header.volunteerSince"
                )}\n${volunteerDate}`}</Text>
              </View>
              {offlineForms ? (
                <Button onPress={upload}>
                  {I18n.t("header.submitOffline")}
                </Button>
              ) : (
                <Button disabled>{I18n.t("header.submitOffline")}</Button>
              )}
              {isSubmitting && <Spinner color="blue" />}
              {isOfflineLoading ? (
                <Spinner color="blue" />
              ) : (
                <Button onPress={cacheOfflineData}>
                  {I18n.t("header.populateOffline")}
                </Button>
              )}
              {!submission && (
                <View>
                  <Text style={styles.calculationText}>
                    {I18n.t("header.failedAttempt")}
                  </Text>
                  <Text style={{ alignSelf: "center" }}>
                    {I18n.t("header.tryAgain")}
                  </Text>
                  <Button onPress={() => setSubmission(null)}>
                    {I18n.t("header.ok")}
                  </Button>
                </View>
              )}
              {submission && (
                <View>
                  <Text style={styles.calculationText}>
                    {I18n.t("header.success")}
                  </Text>
                  <Text style={{ alignSelf: "center" }}>
                    {I18n.t("header.justSubmitted")} {offlineFormCount}{" "}
                    {offlineFormCount > 1 && (
                      <Text>{I18n.t("header.forms")}</Text>
                    )}
                    {offlineFormCount === 1 && (
                      <Text>{I18n.t("header.form")}</Text>
                    )}
                  </Text>
                  <Button onPress={() => setSubmission(null)}>
                    {I18n.t("header.ok")}
                  </Button>
                </View>
              )}
              <View style={{ flexDirection: "row" }}>
                <Button
                  style={styles.calculationTextLeft}
                  onPress={navToSettings}
                >
                  {I18n.t("header.settingsPage")}
                </Button>
                <Button
                  style={styles.calculationTextRight}
                  onPress={navToCounts}
                >
                  {I18n.t("header.surveysCollected")}
                </Button>
              </View>
            </View>
          ) : (
            <FormCounts setShowCounts={setShowCounts} />
          )}
        </View>
      )}
      <IconButton
        size={3}
        style={{
          width: 70,
          backgroundColor: "grey",
          alignSelf: "center",
          opacity: 0.5,
        }}
        onPress={count}
      />
    </View>
  );
};

export default Header;
