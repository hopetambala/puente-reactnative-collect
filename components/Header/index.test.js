import React from 'react';
import renderer from 'react-test-renderer';

import Header from './index';

describe('<Header />', () => {
  it('has 3 children - Unit Testing', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree.children.length).toBe(3);
  });

  it('renders correctly - Snapshot Testing', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
