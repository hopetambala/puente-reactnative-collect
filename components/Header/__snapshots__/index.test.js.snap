// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders correctly - Snapshot Testing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFE680",
      "borderBottomLeftRadius": 20,
      "borderBottomRightRadius": 20,
      "flexDirection": "row",
      "height": 80,
      "justifyContent": "space-evenly",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
        },
        Object {
          "color": "#333",
          "flex": 0.7,
          "fontSize": 30,
          "fontWeight": "bold",
        },
      ]
    }
  >
    PUENTE
  </Text>
  <View
    style={
      Object {
        "borderRadius": 30,
        "color": "#333",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={false}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 22.5,
              "height": 45,
              "width": 45,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={null}
      >
        <Text />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 30,
        "color": "#333",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={false}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 22.5,
              "height": 45,
              "width": 45,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={null}
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
